### PR TITLE
quiz, user course and main tests

### PR DESCRIPTION
**Pull request by EdithHarrison**
**Testing Files:**

- quizController.tests.mjs: Tests for quiz creation, fetching, updating, and deletion.

- userCourseController.tests.mjs: enrolling a user, fetching, and unknown user.

- mainController.tests.mjs: verify the data response.

- Refactor the testing to separate the backend and frontend tests

Testing Instructions:

1. I updated the package.json to have the option to test each file.

2. Before running the tests make sure you have all dependencies.

3. To run all the tests:
`  npm run dev `
 `npm test`

**4. To do individual testings:**

To run the user controller tests:
`  npm run test:user`

To run the frontpage tests:
`npm run dev`
`npm run test:frontpage`

To run the course controller tests:
`npm run test:course`

To run the lesson controller tests:
`npm run test:lesson`

To run the quiz controller tests:
`npm run test:quiz`

To run the user course controller tests:
`npm run test:userCourse`
Screenshot 2024-08-25 at 6 44 08 PM

### CODE REVIEW

File: __tests__/backend/userCourseController.tests.mjs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+// Middleware to simulate authenticated user
+const simulateAuth = (req, res, next) => {
+  req.user = { uid: 'testUserId' }; 
+  next();
+};
+
+// Route setup for user course endpoints
+app.post('/user/enroll', simulateAuth, enrollInCourse);
+app.get('/user/:userId/courses', simulateAuth, getUserCourses);
+
```
Comment (SonaliBedge) [inline __tests__/backend/userCourseController.tests.mjs:18]: There are more endpoints in the userCourseController.mjs file, such as getSuggestedCoursesForUser and getCoursesForUser. You might need to add more test cases for these.

File: __tests__/backend/mainController.tests.mjs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    test('should return the correct JSON response', async () => {
+        const response = await supertest(app)
+            .get('/')
+            .expect(200);
+
+        expect(response.body).toEqual({
+            data: 'This is a full stack app!'
+        });  // Verify that the response contains the correct data
+    });
+});
```
Comment (SonaliBedge) [inline __tests__/backend/mainController.tests.mjs:19]: Do we need to add test file for mainController.mjs file?

File: __tests__/backend/quizController.tests.mjs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    next();
+};
+
+// Route setup for quiz endpoints
+app.post('/course/:courseId/lesson/:lessonId/quiz', simulateAuth, createQuiz);
+app.get('/course/:courseId/lesson/:lessonId/quiz/:quizId', simulateAuth, getQuiz);
+app.get('/course/:courseId/lesson/:lessonId/quizzes', simulateAuth, getAllQuizzesForLesson);
+app.patch('/course/:courseId/lesson/:lessonId/quiz/:quizId', simulateAuth, updateQuiz);
+app.delete('/course/:courseId/lesson/:lessonId/quiz/:quizId', simulateAuth, deleteQuiz);
+
```
Comment (SonaliBedge) [inline __tests__/backend/quizController.tests.mjs:28]: This is good.

File: __tests__/backend/userCourseController.tests.mjs (hunk 1 above)
Comment (EdithHarrison) [reply __tests__/backend/userCourseController.tests.mjs:18]: Same, these new endpoints were not in the main when I was doing the tests. I will address this in the next PR as soon as all the new changes in the main are merged.

File: __tests__/backend/mainController.tests.mjs (hunk 2 above)
Comment (EdithHarrison) [reply __tests__/backend/mainController.tests.mjs:19]: I will remove this test as it is not relevant to our application.

**Review by SonaliBedge — APPROVED** [review]
image

**Review by SonaliBedge — APPROVED** [review]

**Pull request merged**